### PR TITLE
Add JDBC Drivers object to backup/restore.

### PR DESCRIPTION
**Pull request by anilsrivastava1202**
Added JDBC Drivers object to backup/restore.

### CODE REVIEW

File: src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java [hunk 1]
```diff
@@ -15,8 +16,7 @@
 import java.util.Map;
 import java.util.stream.Collectors;
 
-import com.delphix.masking.initializer.maskingApi.endpointCaller.PostMountInformation;
```
[the remark this answers is not in the document]
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java]: It was my bad, seems like IDE did automatically. I will revert it.

File: README.md [hunk 2]
```diff
@@ -205,6 +206,18 @@ the required file format at step up time, the job can be configured properly. Fo
 3. User wants to restore that masking job to another engine. They can achieve this as long as
  `my_file_format.txt`  is next to the backup file, or in the backup folder.
 
+# JDBC Drivers
+
+Unfortunately the masking engine does not provide a way to retrieve JDBC driver files that have been uploaded,
```
[the remark this answers is not in the document]
Comment (anilsrivastava1202) [reply README.md]: Currently, I will prefer the API endpoints to back up/ restore all drivers as API endpoints are already tested. Sync functionality for back up/ restore all drivers is still in development and also need to test properly. But yes we can change it to sync later on once it will be available.

File: src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java [hunk 3]
```diff
@@ -278,6 +257,33 @@ private void handleFileFormats(List<FileFormat> fileFormats) throws IOException
...
+                if (existingJdbcDriver.getJdbcDriverId().equals(jdbcDriver.getJdbcDriverId())) {
+                    apiCallDriver.makePutCall(new PutJdbcDriver(jdbcDriver, driverFile));
+                    isUpdate = true;
+                }
+            }
+            if(!isUpdate) {
+                if(driverFile != null && driverFile.exists() && !driverFile.isDirectory()) {
+                    apiCallDriver.makePostCall(new PostJdbcDriver(jdbcDriver, driverFile));
+                } else {
+                    logger.info(
```
[the remark this answers is not in the document]
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java]: Yes, I think you are right we should fail the whole restore operation as we are doing for FileFormat post-call. I will fix this issue.

File: README.md [hunk 4]
```diff
@@ -205,6 +206,18 @@ the required file format at step up time, the job can be configured properly. Fo
...
+so this tool has no way to retrieve them at backup. However, masking engine does backup of JDBC Drivers MDS information. If a JDBC Driver is backed up, then if the user simply provides
+the required driver zip file at step up time, the JDBC Driver can be configured properly. For example:
+
+1. User sets up a JDBC Driver using driver zip file named `JdbcDriverId.zip`.
+It is important that driver zip file name should be use `JdbcDriverId`, eg. If `JdbcDriverId` is `1`, then driver zip file should be named as `1.zip`.
```
[the remark this answers is not in the document]
Comment (anilsrivastava1202) [reply README.md]: Thanks, I will prefer `JDBC_DRIVER_<DriverId>_<DriverName>.zip` option. I will fix this issue.

[the remark this answers is not in the document]
Comment (anilsrivastava1202) [reply README.md]: `Driver name` can contain spaces and also can be case sensitive so avoided appending driver name with the zip file name. I used the format `JDBC_DRIVER_<ID>`.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java [hunk 5]
```diff
@@ -403,6 +408,39 @@ private void backupUsers() throws ApiCallException {
         }
     }
 
+    private void backupJdbcDrivers() throws ApiCallException {
+        GetSystemInformations getSystemInformations = new GetSystemInformations();
+        apiCallDriver.makeGetCall(getSystemInformations);
+        SystemInformation systemInformation = getSystemInformations.getSystemInformation();
+        String[] versions = systemInformation.getVersion().split("[.]");
+        if(versions.length == 4) {
+            int mejarVersion = 0;
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: `majorVersion`

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java [hunk 6]
```diff
@@ -403,6 +408,39 @@ private void backupUsers() throws ApiCallException {
...
+                mejarVersion = Integer.parseInt(versions[0]);
+            }
+            if(StringUtils.isNumeric(versions[1])) {
+                minorVersion = Integer.parseInt(versions[1]);
+            }
+            if(StringUtils.isNumeric(versions[2])) {
+                microVersion = Integer.parseInt(versions[2]);
+            }
+            // backup JDBC Drivers only if the release version is 6.0.1.0 or higher
+            if(mejarVersion >= 6 && minorVersion >= 0 && microVersion >= 1) {
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Thanks for doing this!

Can we move this logic to another function so that we can just call it from other places as well?

Maybe in the `run` function, we could gather the version info from the engine and then use it in a check version function.
```
if (versionAbove(6, 0, 1)) {
   // backup jdbc driver
}
```

File: generator/src/main/resources/endpointDefinitions/SystemInformation.yaml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+memberVariables:
+  - name: version
+    type: String
+  - name: build
+    type: String
+  - name: latestGlobalKeyResetTime
+    type: String
+  - name: patentNumbers
+    type: List
+    listType: String
```
Comment (bpg130) [inline generator/src/main/resources/endpointDefinitions/SystemInformation.yaml]: I wonder if it makes sense to have this file given that we are not actually using it to generate anything. Do you know if it works fine without this file?

Do you think it makes sense to have this file given it isn't actually doing anything? (I'm not sure if we have ever run into this before so this could set the precedent).

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java (hunk 5 above)
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Thanks.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java (hunk 6 above)
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Done.

Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: I used this file to generate the system-information classes, but later on, I forgot to remove it. Thanks.

File: generator/src/main/resources/endpointDefinitions/SystemInformation.yaml (hunk 7 above)
Comment (anilsrivastava1202) [reply generator/src/main/resources/endpointDefinitions/SystemInformation.yaml]: I used this file to generate the system-information classes, but later on, I forgot to remove it. Thanks.

File: src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java [hunk 8]
```diff
@@ -278,6 +287,26 @@ private void handleFileFormats(List<FileFormat> fileFormats) throws IOException
...
+            boolean isUpdate = false;
+            for (JdbcDriver existingJdbcDriver : getJdbcDriver.getJdbcDrivers()) {
+                if (existingJdbcDriver.getJdbcDriverId().equals(jdbcDriver.getJdbcDriverId())) {
+                    apiCallDriver.makePutCall(new PutJdbcDriver(jdbcDriver, driverFile));
+                    isUpdate = true;
+                }
+            }
+            if(!isUpdate) {
+                apiCallDriver.makePostCall(new PostJdbcDriver(jdbcDriver, driverFile));
+            }
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java]: Using the JDBC driver id as a way to match JDBC drivers actually doesn't work properly across engines. Instead, we should be using the jdbc driver name in the file and the jdbc driver name when comparing existing jdbc drivers.

Lets say I upload an oracle jdbc driver to masking engine `A` and an mssql jdbc driver to masking engine `B`. Both of these jdbc drivers may have an ID of `1`.  If I backup engine `A` and then restore the engine `B`, then I will be overwriting the mssql driver with the oracle driver, even though they are not the same driver.

A similar example is file formats. Two different file formats may have the same id on different engines, so instead of using the file format id, we use the file format name.

```
        // Add any pre-existing file formats to the map so file metadata can be uploaded properly
        GetFileFormats getFileFormats = new GetFileFormats();
        apiCallDriver.makeGetCall(getFileFormats);
        for (FileFormat fileFormat : getFileFormats.getFileFormats()) {
            if (!this.fileFormats.containsKey(fileFormat.getFileFormatName())) {
                this.fileFormats.put(fileFormat.getFileFormatName(), fileFormat.getFileFormatId());
            }
        }
```

Please also update the exected name of the jdbc driver zip file to be `JDBC_DRIVER_<DRIVER_NAME>.zip` instead of `<ID>`.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java [hunk 9]
```diff
@@ -239,6 +247,32 @@ private void finishBackup() {
         }
     }
 
+    private boolean isBackupRequired(int inputMajorVersion, int inputMinorVersion, int inputMicroVersion) {
+        int majorVersion = 0;
+        int minorVersion = 0;
+        int microVersion = 0;
+        GetSystemInformations getSystemInformations = new GetSystemInformations();
+        apiCallDriver.makeGetCall(getSystemInformations);
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Nit: Every time this function is called, we will be making an API call to the engine to fetch the exact same data. Instead, we should cache this value as it will never change for a given run.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java [hunk 10]
```diff
@@ -239,6 +247,32 @@ private void finishBackup() {
         }
     }
 
+    private boolean isBackupRequired(int inputMajorVersion, int inputMinorVersion, int inputMicroVersion) {
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: I'm not sure about the name of this function. What we are really doing is checking if the version of the engine is greater than or equal to that of the inputted version, it doesn't actually have anything to do with backing up, we are just using the result of this check to make that decision.

I suggest changing this to `isEngineVersionAtLeast` or `isEngineVersionGreaterThanOrEqualTo`. This will allow the use of this function to make sense in other contexts as well.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java [hunk 11]
```diff
@@ -239,6 +247,32 @@ private void finishBackup() {
...
+        }
+        if(StringUtils.isNumeric(versions[1])) {
+            minorVersion = Integer.parseInt(versions[1]);
+        }
+        if(StringUtils.isNumeric(versions[2])) {
+            microVersion = Integer.parseInt(versions[2]);
+        }
+        if(majorVersion >= inputMajorVersion && minorVersion >= inputMinorVersion && microVersion >= inputMicroVersion) {
+            return true;
+        }
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: I'm not sure if this version check is correct.

If we need to be on version `6.0.1.0` to run a given backup (like in this change), but the user is version `7.0.0.0`, then I believe this will return false, even though it should return `true` (given that for our case, `7.0.0.0` > `6.0.1.0`.

I believe it would return false because `microVersion >= inputMicroVersion` would not be `true` as `0 >= 1` is `false`, which would cause the whole `if` to be `false`.

We need to say that if the `majorVersion > inputMajorVersion` then we don't care about th minor or micro version and only compare the minor/micro if the versions are equal to each other (and do something similar if the `minorVersion` > `inputMinorVersion` then ignore the micro version).

I suggest that we just use a third-party library to handle this as there are a handful of edge cases. See [this](https://stackoverflow.com/a/6640972) stack overflow answer for an explanation of a good apache library that I have used in the past.

When you fix this, can you also make sure to test this on a trunk developer engine where the version isn't going to be `5.X.X.X` or `6.X.X.X`?

Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Thanks, Your reference to check the version using the third-party library is awesome.  I used `DefaultArtifactVersion`.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java (hunk 10 above)
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Done.

File: src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java (hunk 9 above)
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/BackupDriver.java]: Done.

File: src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java (hunk 8 above)
Comment (anilsrivastava1202) [reply src/main/java/com/delphix/masking/initializer/maskingApi/SetupDriver.java]: Thanks for pointing this issue. Fixed

File: src/main/java/com/delphix/masking/initializer/pojo/SystemInformation.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
+package com.delphix.masking.initializer.pojo;
```
Comment (bpg130) [inline src/main/java/com/delphix/masking/initializer/pojo/SystemInformation.java]: I think that this can be autogenerated?

You should be able to do something like
```java
  - name: profileExpressionNames
    type: list
    listType: String
```
to generate the list.

File: generator/src/main/resources/endpointDefinitions/DatabaseConnector.yaml [hunk 13]
```diff
@@ -36,6 +36,8 @@ memberVariables:
     type: String
   - name: useKerberosAuth
     type: Boolean
+  - name: jdbcDriverId
+    type: Integer
```
Comment (bpg130) [inline generator/src/main/resources/endpointDefinitions/DatabaseConnector.yaml]: I should have caught this before, but we need to store the name of the jdbc driver here as well, not just the id. Then during setup, the name, not the id needs to be used. This is very similar to the file format and file metadata object relationship.

There is no guarantee from engine to engine that the jdbc driver will have the same id, so similar to the reason that we needed to use the name in the zip file, we need to use the name here as well, otherwise, on restore we may end up assigning the wrong jdbc driver to the connector.

There exists a map at `com.delphix.masking.initializer.maskingApi.SetupDriver#fileFormats` that is used to go from file format name -> file format id on the target engine. You will probably need to do something similar.
